### ISSUE TITLE
Raspberry pi zero w with dongle stops reporting / all stop reporting

### ISSUE THREAD

**Issue by JeroenTuinstra** [mKeRix/room-assistant#924]
I am trying for a couple of weeks now to get room presence working. I bought 4 raspberry pi zero w with extra Bluetooth dongles.
Presence has been more than sketchy. The following things constantly happen.

Home Assistant reports either my BLE iphone as away or unavailable
My Apple Watch through Bluetooth Classic the same. 

In my Room assistant companion app they are reported correctly. So why is this information not passed on to Home Assistant. It seems after a reboot that they work for a couple of minutes (maybe 10) and then both entities are reporting wrong again (not in the app, but in Home Assistant). 

I am kindof at a loss what to do. I have replaced one zero with a raspberry pi 3 b+, and am thinking of replacing all the zero's with 3 or 4 pi's. But I am not sure even pi 3 accurately reports the status. 

This is my config file:
```
cluster:
  port: 6415
  peerAddresses:
    - 192.168.10.75:6415
    - 192.168.10.77:6415
    - 192.168.10.79:6415
    - 192.168.10.80:6415
    - 192.168.10.81:6415
  autoDiscovery: true
  weight: 30
global:
  instanceName: bedroom
  integrations:
    - homeAssistant
    - bluetoothLowEnergy
    - bluetoothClassic
homeAssistant:
  mqttUrl: 'mqtt://mqttaddress:1883'
  mqttOptions:
    username: loginname
    password: password
bluetoothLowEnergy:
  hciDeviceId: 0
  allowlist:
    - 1E808694-68E0-433E-A010-BC850C8481F0
    - 4A89AB27-F043-4A8F-B357-5FA63302DC36
  maxDistance: 5
  tagOverrides:
    1E808694-68E0-433E-A010-BC850C8481F0:
      name: Iphone Jeroen BLE
    4A89AB27-F043-4A8F-B357-5FA63302DC36:
      name: Iphone Isaac BLE
bluetoothClassic:
  hciDeviceId: 1
  minRssi: -5
  addresses:
    - '64:d2:c4:7a:72:c2'
  entityOverrides:
    '64:d2:c4:7a:72:c2':
      name: Apple Watch Jeroen BT
```

On 2 of the raspberry pi zero's I constantly get this error when running room-assistant in verbal mode: 
```
BluetoothService: Command failed: hcitool -i hci1 cc "64:d2:c4:7a:72:c2" && hcitool -i hci1 rssi "64:d2:c4:7a:72:c2"
Can't create connection: Input/output error
Get connection info failed: No such file or directory
```
Connection to MQTT broker is fine. New instances are reported to Home Assistant, just their status not.

**Comment by JeroenTuinstra** [comment]
I seem to have narrowed down the problem. I have now installed all but one nodes on either pi4 or pi3B+.
They all have an extra USB bluetooth dongle and in their configuration files the Bluetooth Low Energy is assigned to the USB device and the Bluetooth Classic to the UART device. 
On all I regularly get this error sequence and the device on the Bluetooth Classic (i.e. the Apple Watch) is not registering: 

```
BluetoothService: Locking adapter 1
27/09/2021, 22:47:34 - debug - BluetoothService: Querying for RSSI of 64:d2:c4:7a:72:c2 using hcitool
27/09/2021, 22:47:39 - debug - BluetoothService: Command failed: hcitool -i hci1 cc "64:d2:c4:7a:72:c2" && hcitool -i hci1 rssi "64:d2:c4:7a:72:c2"
Can't create connection: Input/output error
Get connection info failed: No such file or directory

27/09/2021, 22:47:39 - debug - BluetoothService: Unlocking adapter 1
```

**Comment by JeroenTuinstra** [comment]
@Chaphasilor I gave up with Bluetooth Low Energy. The app simply doesn't work and lets the pi's crash regularly. And apparently the developer of the app doesn't care, as he no longe seems to respond to any issues regarding Bluetooth Low Energy. I am just using Bluetooth Classic. The system is running stable now. 

I understand that the input/output error also has to do with the bluetooth device not being present. So it checks on the BT MAC address and if the device is not in range and the cluster doesn't report it properly (cluster doesn't seem to work as it should be either), then the input/output error seems to occur.

**Comment by Chaphasilor** [comment]
@JeroenTuinstra well I'm trying to get Bluetooth Classic to work, thats where I get the error 😅 
It worked fine for a couple of days, then I had the issue, which I fixed by resetting the `hci0` device before restarting `room-assistant`.  
That worked for a couple of days, but now it broke down as well, and strangely on multiple devices at once.

I can't find a valid explanation for it, let alone a fix ://

**Comment by JeroenTuinstra** [comment]
@Chaphasilor What made it work for me was to add an extra BT dongle to the raspberry pi. The extra dongle is recognized and configured automatically to `hci0`. It hasn't stopped working for 2 weeks now.
Before, just using the internal BT, made it crash very often, with the extra dongle, no more.
For BT Low Energy that didn't work for me. Tried to assign low energy to the dongle and classic to internal, but it didn't help.

**Comment by JeroenTuinstra** [comment]
@PeteBa Thank you for this information. I am going to try this and also check what version I have, but I think I have the latest. I have here a RPi Zero W, which I can configure exclusively for BLE, to see if it is stable. I will keep you posted.

**Comment by JeroenTuinstra** [comment]
So far it seems stable on a RPi Zero W with only BLE enabled and the `instanceBeaconEnabled: false` option in the config. The presence is stable. I will implement it now at my home to see if the combination BLE and Classic (on a dongle) will be stable too.

**Comment by PeteBa** [comment]
@JeroenTuinstra , thanks for the update. Let us know how you get on with adding in Classic.

**Comment by JeroenTuinstra** [comment]
So I have enabled both BLE and Classic, with Classic on the dongle. So far it seems to go well ... even though 30 minutes ago BLE room presence said `not_home`, while I was 0.5 meter away from the RPi in het offce. Only when I opened the room assistant app did it update the correct room.
I will keep monitoring it, but the stability of the app was a serious issue in the past. 

At this moment I am **not** using `timeout: 60` in the config, perhaps that has something to do with it.

**Comment by JeroenTuinstra** [comment]
After monitoring today the room presence. I have to conclude that the app is not refreshing in the background. I will indicate the proper room presence for awhile and then it will show not home. When I open the app and try to refresh is updates the right location. 

So sorry to say the app is just not working/reliable. I think I will refer back to Classic.

**Comment by PeteBa** [comment]
Hmm, that's unfortunate. For me the App is pretty reliable. I do experience "updates stopping" but only rarely, say, once every few weeks.

Out of interest, when you say that the "app is not refreshing in the background" do you mean that the updates stop: 1) immediately upon leaving the app using the home button; 2) immediately upon the iphone's screen saver/lock screen activating, or 3) some time later after a random period of time ?

I've taken a look at the iphone app code and it looks like a couple of setting impact advertising behaviour. The first is within the room-assistant app and is the "Auto-Toggle Visibility" that I have as off. Second is the iPhone's "Settings -> Privacy -> Location Services -> Room-Assistant" that I have as Never. Are your equivalents different ?

**Comment by JeroenTuinstra** [comment]
I have set the settings as you said. Which seem counterintuitive, but it didn’t improve. 
So as long as I keep the app open it reports the right room. When I quit the app it switches almost immediately to “not_home”. 
Screen lock doesn’t seem to have any influence. Closing the app does.

**Comment by JeroenTuinstra** [comment]
Having monitored the room presence for today, the following is happening. When the app is opened the room presence is correctly registered. After closing the app it keeps registering the room properly and then after 10 - 15 minutes it puts it on "not_home" untill the app is opened. 
I have set `instanceBeaconEnabled: true` or false, to no difference. With the other settings it doesn't make a difference either. The app or room-assistant simply stops reporting the correct room after 10-15 minutes after the app is closed. Keeping the app open - keeps giving the right room status - which makes me conclude that the error is with the app. 
Basically background reporting is not working correctly. 

I am willing to do more testing but some of my automations are relying on room presence - which is not working/unreliable at the moment.

**Comment by PeteBa** [comment]
I have been checking this out too...

Yes the settings were counter-intuitive. Basically, there is some code in the App that can turn off advertising under certain circumstances (app code [here ](https://github.com/mKeRix/room-assistant-companion-ios/blob/6dc20b170fcc8b3f0773b0663cd47065e17a30d0/room-assistant%20companion/Services/BluetoothManager.swift#L65) and [here](https://github.com/mKeRix/room-assistant-companion-ios/blob/6dc20b170fcc8b3f0773b0663cd47065e17a30d0/room-assistant%20companion/Services/BluetoothManager.swift#L162)) This is part of room assistant's "instanceBeacon" feature that has had some reliability issues. The settings above avoid that code and I was just looking to rule it out of the equation.

When you say "quit the app" above I assume you mean "background" the app by returning to the home screen rather than "force exit" the app via the task switcher ? If you "force exit" the app then updates will stop the next time the iphone changes its mac address (i.e. after ~20 mins). This could be why you see the "not_home" after 10 - 15 minutes.

There are some circumstances where iOS will terminate a background app (e.g. as a result of low memory or high battery consumption). In some cases, iOS will restore the application when the load drops and hence room-assistant can continue on. However, there are cases where iOS does not allow the background bluetooth app to continue (see [here](https://developer.apple.com/library/archive/qa/qa1962/_index.html)). Those scenarios with a NO in the relaunch column will ultimately result in room-assistant loosing the association between the iphone's mac address and the room-assistant ID.

One way to identify this is by looking in your room-assistant logs. You should be able to get to them on your pi with the following command `sudo journalctl -u room-assistant`. If you look at the period just before/after a "not_home" event then the tell-tale sign is if you see a bunch of failed discovery attempts i.e. you only see the first message below and not the corresponding second message.

```
07/11/2021, 18:34:57 - info - BluetoothLowEnergyService: Attempting app discovery for tag 64d09e1ba5b8
07/11/2021, 18:34:59 - info - BluetoothLowEnergyService: Discovered companion app with ID F5976501-5555-40DD-BD46-7E4528619036 for tag 64d09e1ba5b8
```

This would help confirm if your experience is down to the iOS background policy.

**Comment by JeroenTuinstra** [comment]
Thanks for this follow up. When I talk about quitting the application, I mean force quitting the app. I understand now that the app has to be open all the time. I assumed like the mail app and many other apps, it would refresh in the “background” - as in just like the mail app, Facebook, etc, still partially updates even if the application is force quitted.
Going back to the Home Screen is keeping applications opens. The task switcher is according to Apple themselves the possibility to see all your open apps. 
So it is correct that the app only keeps working when it is open, but as I understand now this is by design. The app cannot be exited.  Which is really a shame - as I don’t like having all types of apps open. 
I’ll monitor whether the app keeps updating if keep it open all the time.

**Comment by JeroenTuinstra** [comment]
I checked the app refresh in the background. The headline is on, however room assistant is not listed. I guess it is not capable of doing background refreshing.

**Comment by PeteBa** [comment]
Yep, the app needs to be open and in the background (i.e. in the task switcher list) so that whenever the iPhone's MAC address changes, the RPi room-assistant service can still associate the updates to your iPhone. This happens every twenty-ish minutes so hopefully you now see the updates continue beyond that point. 

The room-assistant app does refresh in the background, it is just that bluetooth apps are granted that right automatically and so they dont need a toggle switch in that Settings menu.

**Comment by JeroenTuinstra** [comment]
@PeteBa I guess we can discuss the definition of "background", but really the task switcher on the iphone switches between open apps, not background apps. For example the mail app doesn't have to be open in order to receive and notify you of new emails. Almost no app needs to remain open (or in the task switcher) in order to update.
This app would be great if it could properly work in the background and not rely on you keeping it in the task switcher. But indeed when you keep the app in the task switcher it updates continuously, so that works.
If the app is ever rewritten to work really in the background, then I will give BLE with the iphone another shot. Till then, Classic is I guess the way.
